src/server/routes/register.ts

```diff
@@ -110,6 +133,262 @@ router.post(
+					timeUntilTokenExpiry: encryptedState.stateHandleExpiresAt
+						? new Date(encryptedState.stateHandleExpiresAt).getTime() -
+							Date.now()
+						: undefined,
+				},
+			}),
+			pageTitle: 'Check Your Inbox',
+		});
+		return res.type('html').send(html);
+	}
+	return res.redirect(
+		303,
+		addQueryParamsToPath('/register/email', state.queryParams),
+	);
+});
+
+// Handler for the passcode input form
+router.post(
+	'/register/code',
+	redirectIfLoggedIn,
+	handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+		const { requestId } = res.locals;
+		const { code } = req.body;
+
+		const encryptedState = readEncryptedStateCookie(req);
+
+		// make sure we have the encrypted state cookie and the code otherwise redirect to the email registration page
+		if (encryptedState && encryptedState.stateHandle && code) {
+			const { stateHandle } = encryptedState;
+
+			try {
+				// check the state handle is valid and we can proceed with the registration
+				const introspectResponse = await introspect(
+					{
+						stateHandle,
+					},
+					requestId,
+				);
+
+				// check if the remediation array contains a "enroll-authenticator"	object
+				// if it does, then we know the stateHandle is valid and we're in the correct state
+				validateIntrospectRemediation(
+					introspectResponse,
+					'enroll-authenticator',
+				);
+
+				// attempt to answer the passcode challenge, if this fails, it falls through to the catch block where we handle the error
+				const challengeAnswerResponse = await challengeAnswerPasscode(
+					stateHandle,
+					{ passcode: code },
+					requestId,
+				);
+
+				// if passcode challenge is successful, we can proceed to the next step
+				// which is to enroll a password authenticator, as we still need users to set a password for the time being
+				// we first look for the password authenticator id deep in the response
+				const passwordAuthenticatorId =
+					challengeAnswerResponse.remediation.value
+						.flatMap((remediation) => {
+							if (remediation.name === 'select-authenticator-enroll') {
+								const parsedRemediation =
+									selectAuthenticationEnrollSchema.safeParse(remediation);
+
+								if (parsedRemediation.success) {
+									return parsedRemediation.data.value.flatMap((value) => {
+										if (value.name === 'authenticator') {
+											return value.options.flatMap((option) => {
+												if (option.label === 'Password') {
+													if (
+														option.value.form.value.some(
+															(v) => v.value === 'password',
+														)
+													) {
+														return [
+															option.value.form.value.find(
+																(v) => v.name === 'id',
+															)?.value,
+														];
+													}
+												}
+											});
+										}
+									});
+								}
+							}
+						})
+						.filter(
+							(id): id is string => typeof id === 'string' && id.length > 0,
+						);
```

Comment:
This is terrifying. No suggestions to fix it, it just is. :(

Reply:
A further thought: is there any way at all that we could leverage zod to handle this complex nested object? It feels like a maintenance nightmare otherwise.

Reply:
this is leveraging Zod as best I can 😱

it's just annoying at there are so many states to what the IDX API returns, it's not consistent, so we always have to double check everything.